src/main/java/com/gargoylesoftware/htmlunit/WebClient.java

```diff
@@ -1681,6 +1688,97 @@ else if (status == HttpStatus.SC_TEMPORARY_REDIRECT
         return webResponse;
     }
 
+    /**
+     * Validate the cached {@link WebResponse} to see if it could be used
+     * @param fromCache the {@link WebResponse} stored in cache that need to be validated before using
+     * @param webRequest the {@link WebRequest} that go with the {@link WebResponse} from cache
+     */
+    private WebResponse validateWebResponseFromCache(final WebResponse fromCache, final WebRequest webRequest) throws IOException {
+        // implementation based on rfc9111 https://www.rfc-editor.org/rfc/rfc9111#name-validation
+        if (HeaderUtils.containsETag(fromCache)) {
+            webRequest.setAdditionalHeader(HttpHeader.IF_NONE_MATCH, fromCache.getResponseHeaderValue(HttpHeader.ETAG));
+        }
+        if (HeaderUtils.containsLastModified(fromCache)) {
+            webRequest.setAdditionalHeader(HttpHeader.IF_MODIFIED_SINCE, fromCache.getResponseHeaderValue(HttpHeader.LAST_MODIFIED));
+        }
+
+        WebResponse webResponse = getWebConnection().getResponse(webRequest);
+
+        if (webResponse.getStatusCode() >= HttpStatus.SC_INTERNAL_SERVER_ERROR) {
+            return new WebResponseFromCache(fromCache, webRequest);
+        }
+
+        if (webResponse.getStatusCode() == HttpStatus.SC_NOT_MODIFIED) {
+            Set<NameValuePair> alteredHeaders = new HashSet<>(fromCache.getResponseHeaders());
```

Comment:
this breaks ordering
```suggestion
            Set<NameValuePair> alteredHeaders = new LinkedHashSet<>(fromCache.getResponseHeaders());
```